File: backend/src/services/csvService.js
const { parse } = require('csv-parse');
const Class = require('../models/Class');

/**
 * Parse a CSV Buffer and return an array of plain objects.
 * Expects headers: first_name, last_name, email, class_label, year, photo_url
 */
function parseCsv(buffer) {
  return new Promise((resolve, reject) => {
    parse(buffer, { columns: true, skip_empty_lines: true, trim: true }, (err, records) => {
      if (err) return reject(err);
      resolve(records);
    });
  });
}

/**
 * Validate records and resolve class IDs.
 * Returns { students: [...], errors: [...] }
 */
async function processCsvRecords(records) {
  const students = [];
  const errors = [];

  // Cache classes to avoid repeated DB queries
  let classCache = null;
  const getClasses = async () => {
    if (!classCache) classCache = await Class.findAll();
    return classCache;
  };

  for (let i = 0; i < records.length; i++) {
    const rec = records[i];
    const rowNum = i + 2; // row 1 = header

    if (!rec.first_name || !rec.last_name) {
      errors.push({ row: rowNum, error: 'first_name and last_name are required' });
      continue;
    }

    let class_id = null;
    if (rec.class_label) {
      const classes = await getClasses();
      const existing = classes.find((c) => c.label === rec.class_label);
      if (existing) {
        class_id = existing.id;
      } else {
        try {
          const created = await Class.create({ label: rec.class_label, year: rec.year || null });
          class_id = created.id;
          classCache = null; // invalidate cache
        } catch (err) {
          errors.push({ row: rowNum, error: `Could not create class "${rec.class_label}": ${err.message}` });
        }
      }
    }

    students.push({
      first_name: rec.first_name,
      last_name: rec.last_name,
      email: rec.email || null,
      class_id,
      photo_url: rec.photo_url || null,
    });
  }

  return { students, errors };
}

module.exports = { parseCsv, processCsvRecords };

File: backend/src/services/csv_service.py
import csv
import io

from ..models.class_ import ClassModel


def parse_csv(content: bytes) -> list[dict]:
    text = content.decode("utf-8-sig")
    reader = csv.DictReader(io.StringIO(text))
    return [row for row in reader]


def process_csv_records(records: list[dict]):
    students: list[dict] = []
    errors: list[dict] = []
    class_cache: list[dict] | None = None

    def get_classes() -> list[dict]:
        nonlocal class_cache
        if class_cache is None:
            class_cache = ClassModel.find_all()
        return class_cache

    for i, rec in enumerate(records):
        row_num = i + 2
        first_name = (rec.get("first_name") or "").strip()
        last_name = (rec.get("last_name") or "").strip()

        if not first_name or not last_name:
            errors.append({"row": row_num, "error": "first_name and last_name are required"})
            continue

        class_id = None
        class_label = (rec.get("class_label") or "").strip()
        if class_label:
            classes = get_classes()
            existing = next((c for c in classes if c["label"] == class_label), None)
            if existing:
                class_id = existing["id"]
            else:
                try:
                    year_value = (rec.get("year") or "").strip() or None
                    created = ClassModel.create(
                        label=class_label,
                        year=year_value,
                    )
                    class_id = created["id"]
                    class_cache = None
                except Exception as e:
                    errors.append(
                        {"row": row_num, "error": f'Could not create class "{class_label}": {e}'}
                    )

        students.append(
            {
                "first_name": first_name,
                "last_name": last_name,
                "email": rec.get("email") or None,
                "class_id": class_id,
                "photo_url": rec.get("photo_url") or None,
            }
        )

    return students, errors
